daemon/mgr/container_list.go

```diff
@@ -0,0 +1,211 @@
+	if option == nil {
+		return &filterContext{}, nil
+	}
+
+	// validate filter here
+	if err := opts.ValidateFilter(option.Filter); err != nil {
+		return nil, err
+	}
+	return &filterContext{
+		condition:  option.Filter,
+		all:        option.All,
+		filterFunc: option.FilterFunc,
+	}, nil
+}
+
+// matchFilter filters value matchs field of condition.
+func (fc *filterContext) matchFilter(field, value string) (bool, bool) {
+	filters, exist := fc.condition[field]
+	if !exist {
+		// jump filter if field is not exist
+		return true, false
+	}
+
+	if value == "" || len(filters) == 0 {
+		return false, false
+	}
+
+	var (
+		match = false
+		err   error
+	)
+	for _, f := range filters {
+		if match, err = regexp.MatchString(f, value); err != nil {
+			continue
+		}
+		if match {
+			break
+		}
+	}
+
+	return false, match
+}
+
+// matchKVFilter filters map value matchs field of condition, also support
+// filter not equal, as for unequal condition `key=value`, we filter `$k=$v`
+// which $k equal key and $v not equal value.
+func (fc *filterContext) matchKVFilter(field string, value map[string]string) (bool, bool) {
+	filters, exist := fc.condition[field]
+	if !exist {
+		// jump filter if field is not exist
+		return true, false
+	}
+
+	if len(filters) == 0 {
+		return false, false
+	}
+
+	// repeat value not support, eg: {"label": []string{"a=b", "a=c"}}, c will overwrite b
+	equalKV := make(map[string]string)
+	unequalKV := make(map[string]string)
+	for _, f := range filters {
+		if f == "" {
+			continue
+		}
+		splits := strings.SplitN(f, "!=", 2)
+		if len(splits) == 2 {
+			unequalKV[splits[0]] = splits[1]
+			continue
+		}
+
+		splits = strings.SplitN(f, "=", 2)
+		if len(splits) == 2 {
+			equalKV[splits[0]] = splits[1]
+			continue
+		}
+
+		equalKV[f] = ""
+	}
+
+	if len(equalKV) == 0 && len(unequalKV) == 0 {
+		return false, false
+	}
+
+	match := false
+	for k, v := range value {
+		// filter equal condition
+		if equalValue, exist := equalKV[k]; exist {
+			if equalValue == "" {
+				match = true
+				break
+			}
+			if equalValue == v {
+				match = true
+				break
+			}
+		}
+		// filter unequal condition
+		if unequalValue, exist := unequalKV[k]; exist && unequalValue != v {
+			match = true
+			break
+		}
+	}
+
+	return false, match
+}
+
+// filter does all select container work.
+func (fc *filterContext) filter(c *Container) bool {
+	isNonStop := c.IsRunningOrPaused()
+
+	// if filterFunc is defined, we not try other filter condition
+	if fc.filterFunc != nil {
+		if match := fc.filterFunc(c); match {
+			// consider if need to add non-running container
+			return isNonStop || fc.all
+		}
+		return false
+	}
+
+	if len(fc.condition) == 0 {
+		// if no filter condition, skip filter process,
+		// add container according to flag `all` or running status
+		return isNonStop || fc.all
+	}
+
+	// filter must meets all condition
+	var match, jump, statusKey = true, true, false
+	for name := range fc.condition {
```

Comment:
I have idea about this part.

The `jump` boolean means we should start next round check, right?  the `matchKVFilter` will return one boolean value if we run the `filters, exist := fc.condition[field]` before check.  Therefore, we don't need to check the `jump` in the following condition. 

Does it make senses to you?

Reply:
I also think about this, `filters, exist := fc.condition[field]` if field not exist, whether filter function return true or false is not make sense. Maybe need more think.